animation_nodes/nodes/mesh/generation/line.py

```diff
@@ -1,18 +1,52 @@
 import bpy
+from bpy.props import *
+from .... data_structures import Mesh
 from .... base_types import AnimationNode
-from .... algorithms.mesh_generation.line import getLineMesh
+from .... algorithms.mesh_generation.line import getLineMesh, edges
+
+lineModeItems = [
+    ("STARTEND", "Start-End", "Line from start-end", "NONE", 0),    
+    ("POINTS", "Points", "Line from vector list", "NONE", 1)
+]
 
 class LineMeshNode(bpy.types.Node, AnimationNode):
     bl_idname = "an_LineMeshNode"
     bl_label = "Line Mesh"
 
+    lineMode: EnumProperty(name = "Line Mode", default = "STARTEND",
+        items = lineModeItems, update = AnimationNode.refresh)
+
     def create(self):
-        self.newInput("Vector", "Start", "start")
-        self.newInput("Vector", "End", "end", value = [5, 0, 0])
-        self.newInput("Integer", "Steps", "steps", value = 2, minValue = 2)
+        if self.lineMode == "STARTEND":
+            self.newInput("Vector", "Start", "start")
+            self.newInput("Vector", "End", "end", value = [5, 0, 0])
+            self.newInput("Integer", "Steps", "steps", value = 2, minValue = 2)
+        elif self.lineMode == "POINTS":
+            self.newInput("Vector List", "Points", "points")
+            self.newInput("Boolean", "Cyclic", "cyclic", value = False)
 
         self.newOutput("Mesh", "Mesh", "mesh")
 
-    def execute(self, start, end, steps):
+    def draw(self, layout):
+        layout.prop(self, "lineMode", text = "")
+
+    def getExecutionFunctionName(self):
+        if self.lineMode == "STARTEND":
+            return "execute_StartEndLine"
+        elif self.lineMode == "POINTS":
+            return "execute_PointsLine"
+
+    def execute_StartEndLine(self, start, end, steps):
         steps = max(steps, 2)
-        return getLineMesh(start, end, steps)
\ No newline at end of file
+        return getLineMesh(start, end, steps)
+
+    def execute_PointsLine(self, points, cyclic):
+        if len(points) < 2: return Mesh()
+        pointCount = len(points)
+        edgeIndices = edges(pointCount)
+        if cyclic:
+            edgeIndices.append((pointCount - 1, 0))
+            return Mesh(points, edgeIndices, skipValidation = True)
+        else:    
+            return Mesh(points, edgeIndices, skipValidation = True)
```

Comment:
I think we should create a function for it in `line.pyx` instead of doing things manually here.

Reply:
Okay. Omar, what type of identifier for boolean in Cython which is cyclic in this case.

Reply:
I'm using this `from cpython cimport bool`.

Reply:
Here is the function for Lines mesh,
```
def getLinesMesh(Vector3DList points, bool cyclic):
    cdef long pointCount = len(points)
    edgeIndices = edges(pointCount)
    if cyclic:
        edgeIndices.append((pointCount - 1, 0))
        return Mesh(points, edgeIndices, skipValidation = True)
    else:    
        return Mesh(points, edgeIndices, skipValidation = True)
```

Reply:
Execute `cython -a line.pyx`. A `line.html` file will be generated. This is an annotated version of the code, yellow lines means a kind of interaction with Python, we try to minimize that as much as possible. So, the less yellow you have, the better. In this code, the third line will be yellow, because edgeIndices is not typed. So will the fifth line, because you are calling append.

Reply:
I have updated the code,
```
def getLinesMesh(Vector3DList points, bool cyclic):
    cdef long pointCount = len(points)
    cdef EdgeIndicesList edgeIndices = edges(pointCount)
    if cyclic:
        edgeIndices.append((pointCount - 1, 0))
        return Mesh(points, edgeIndices, skipValidation = True)
    else:    
        return Mesh(points, edgeIndices, skipValidation = True)
```

Reply:
I would also try to avoid calling append.

Reply:
Sorry, but I do know how I can add an extra edge-indices in EdgeIndicesList.

Reply:
Pass `cyclic` to the `edges` function and generate the edges accordingly.

Reply:
Thanks. Here is the code,
```
# Edges
############################################

def edges(long steps, bool cyclic):
    cdef long edgeAmount
    if cyclic:
        edgeAmount = steps
    else:
        edgeAmount = steps - 1
    cdef EdgeIndicesList edges = EdgeIndicesList(length = edgeAmount)
    cdef long i 
    for i in range(edgeAmount):
        edges.data[i].v1 = i
        edges.data[i].v2 = i + 1
    if cyclic:
        edges.data[edgeAmount - 1].v2 = 0
    return edges


# Points
############################################

def getLinesMesh(Vector3DList points, bool cyclic):
    cdef long pointCount = len(points)
    cdef EdgeIndicesList edgeIndices = edges(pointCount, cyclic)
    return Mesh(points, edgeIndices, skipValidation = True)
```

Reply:
Use `bint` for boolean types.

Reply:
Okay.